index/segment/mem/segment.go

```diff
@@ -93,26 +97,60 @@ func NewSegment(offset postings.ID, opts Options) (sgmt.MutableSegment, error) {
 
 func (s *segment) Insert(d doc.Document) error {
 	s.state.RLock()
+	defer s.state.RUnlock()
 	if s.state.closed {
-		s.state.RUnlock()
 		return sgmt.ErrClosed
 	}
 
 	if s.state.sealed {
-		s.state.RUnlock()
 		return errSegmentSealed
 	}
 
-	// Validate that the document contains only valid UTF-8.
+	// Check if the document contains a document ID and ensure it's fields are valid UTF-8.
+	var docID []byte
 	for _, f := range d.Fields {
+		// TODO: Should we enforce uniqueness of field names?
 		if !utf8.Valid(f.Name) {
 			return fmt.Errorf("document contains invalid field name: %v", f.Name)
 		}
+
+		if bytes.Equal(f.Name, doc.IDFieldName) {
+			if docID != nil {
+				return errMultipleIDs
+			}
+
+			if len(docID) == 0 {
+				return errZeroLengthID
+			}
+
+			docID = f.Value
+		}
+
 		if !utf8.Valid(f.Value) {
 			return fmt.Errorf("document contains invalid field value: %v", f.Value)
 		}
 	}
 
+	if docID != nil {
+		pl, err := s.matchTerm(doc.IDFieldName, docID)
+		if err != nil {
+			return fmt.Errorf("could not determine if index already contains document ID '%s'", docID)
+		}
+		if !pl.IsEmpty() {
+			// Index already contains this document.
+			return nil
+		}
+	} else {
+		uuid, err := util.NewUUID()
```

Comment:
I can see the value behind this for certain use-cases. Mind hiding this behind an Option?